### PR TITLE
Introduce the KVDB to phala runtime

### PR DESCRIPTION
**Pull request by g302ge**
This PR is targetd to replace phactory and pink MemoryDB into underlying KVDB.

Currenly, Pink and Phactory both use the MemoryDB, Morever after a specificed interval, Phactory will take a checkpoint not only over it's RuntimeState but the Pink storage. Unfortunatly, just replace the Pink storage could cause the whole system into inconsistence. For example:

Assume that, Phactory is now sync `100th` block from pherry and Pink is working on `99th` block, if we just simply commit the transaction in Pink into the underlying storage, when the phactory crash now, Phactory checkpoint maybe at `59th` block. When phala restart, from `59th` block to `99` block storage changes is duplicated, 

# Implementation Details 

firstly we should introduce the KVDB as a crate to support base semantics used for both `Pink` and `Phactory`. 

`Pink` is just need a `TrieBackend`. So simplly wraps the underlying storage as `TrieBackendStorage` makes thing sound. And for `Phactory` it just needs the `RawDB` interface not `HashDB`. Due to this, we should introduce another interface over pkvdb to supply the `RawDB` interface.

And then, commit all transaction both `Pink` changes and `Phactory` checkpoint per block makes thing sound.

### CODE REVIEW

File: crates/pkvdb/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+pub(crate) struct Inner<T>(pub(crate) T, pub(crate) i32);
+
+impl<T> From<Vec<u8>> for Inner<T>
+where
+    T: for<'a> From<&'a [u8]>,
+{
+    fn from(v: Vec<u8>) -> Self {
+        let (rc, val) = v.split_at(4);
+        let value = T::from(val);
```
Comment (kvinwang) [inline crates/pkvdb/src/lib.rs:47]: Still double allocation

File: crates/pkvdb/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+const SYNC_TRIE_ROOT_STORAGE_KEY: &str = "phaladb_sync_storage_key";
+
+// PhalaDB used for keep tract root and other memory
+// NOTE: you have to change the root at every checkpoint and save to transaction to
+// make the DB work correctly
+pub struct PhalaDB<H: hash_db::Hasher> {
+    db: Arc<Mutex<DB>>,
+    pink_overlay: Arc<Mutex<MemoryDB<H>>>,
+    sync_overlay: Arc<Mutex<MemoryDB<H>>>,
+    raw_map: BTreeMap<Vec<u8>, Vec<u8>>,
```
Comment (kvinwang) [inline crates/pkvdb/src/lib.rs:77]: Only one overlay(the raw_map) here should be enough. pink/sync are user logic.

File: crates/pkvdb/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    pub pink_root: Option<H::Out>,
+    pub sync_root: Option<H::Out>,
```
Comment (kvinwang) [inline crates/pkvdb/src/lib.rs:79]: The root should be tracked by user, don't put them in this low level DB layer.

File: crates/pkvdb/src/trie.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    fn as_plain_db(&self) -> &dyn PlainDB<H::Out, DBValue> {
+        self
+    }
+
+    fn as_plain_db_mut(&mut self) -> &mut dyn PlainDB<H::Out, DBValue> {
+        self
+    }
+}
+
+impl<H: Hasher> HashDB<H, DBValue> for TrieEssenceStorage<H>
```
Comment (kvinwang) [inline crates/pkvdb/src/trie.rs:131]: So many `impl HashDB for ...`. Can we reduce some of them?

**Comment by h4x3rotab** [comment]
Closed. The work is continued in #928

**Pull request closed without merging**